### PR TITLE
Add new trainer for GAN-based training

### PR DESCRIPTION
**Pull request by kan-bayashi**
Subset of #3423

This PR adds new trainer for GAN-based training.
To do this, a new abstract class `AbsGANESPnetModel` is added.
To use this trainer, the model must be accept `forward_generator=True` as an argument,
which switch the loss calculation of generator and discriminator.

```python
        >>> from espnet2.tasks.abs_task import AbsTask
        >>> class YourESPnetModel(AbsGANESPnetModel):
        ...     def forward(self, input, input_lengths, forward_generator=True):
        ...         ...
        ...         if forward_generator:
        ...             # return loss for the generator
        ...             # optim idx 0 indicates generator optimizer
        ...             return dict(loss=loss, stats=stats, weight=weight, optim_idx=0)
        ...         else:
        ...             # return loss for the discriminator
        ...             # optim idx 1 indicates discriminator optimizer
        ...             return dict(loss=loss, stats=stats, weight=weight, optim_idx=1)
        >>> class YourTask(AbsTask):
        ...     @classmethod
        ...     def build_model(cls, args: argparse.Namespace) -> YourESPnetModel:
```

### CODE REVIEW

File: espnet2/train/gan_trainer.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+    @torch.no_grad()
+    def plot_attention(
+        cls,
+        model: torch.nn.Module,
+        output_dir: Optional[Path],
+        summary_writer: Optional[SummaryWriter],
+        iterator: Iterable[Tuple[List[str], Dict[str, torch.Tensor]]],
+        reporter: SubReporter,
+        options: TrainerOptions,
+    ) -> None:
```
Comment (takenori-y) [inline espnet2/train/gan_trainer.py:405]: Can we call here `plot_attention` method of the parent instead of writing the copy?

File: espnet2/train/abs_gan_espnet_model.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        ...     @classmethod
+        ...     def build_model(cls, args: argparse.Namespace) -> YourESPnetModel:
+    """
+
+    @abstractmethod
+    def forward(
+        self,
+        forward_generator: bool,
+        **batch: torch.Tensor,
+    ) -> Dict[str, Union[torch.Tensor, Dict[str, torch.Tensor], int]]:
```
Comment (takenori-y) [inline espnet2/train/abs_gan_espnet_model.py:48]: Could you add the comments about `Returns` to let know us the meanings of the returns? (`forward` in `trainer.py` has also no comments).

File: espnet2/train/gan_trainer.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    @classmethod
+    def build_options(cls, args: argparse.Namespace) -> TrainerOptions:
+        """Build options consumed by train(), eval(), and plot_attention()"""
+        assert check_argument_types()
+        return build_dataclass(TrainerOptions, args)
+
+    @classmethod
+    def add_arguments(cls, parser: argparse.ArgumentParser):
+        parser.add_argument(
+            "--generator_first",
```
Comment (takenori-y) [inline espnet2/train/gan_trainer.py:105]: Different way: `roles = ["generator", "discriminator"]`.
I know this is redundant but it can handle more than two models. I’ll leave it up to you.

File: espnet2/train/gan_trainer.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+                    # Unscales the gradients of optimizer's assigned params in-place
+                    for iopt, optimizer in enumerate(optimizers):
+                        if optim_idx is not None and iopt != optim_idx:
+                            continue
+                        scaler.unscale_(optimizer)
+
+                # gradient noise injection
+                if grad_noise:
+                    raise NotImplementedError(
+                        "Gradient noise injection is not supported."
```
Comment (takenori-y) [inline espnet2/train/gan_trainer.py:252]:
```suggestion
                        "Gradient noise injection is not supported in GAN-based training."
```

File: espnet2/train/gan_trainer.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+                torch.distributed.all_reduce(iterator_stop, ReduceOp.SUM)
+                if iterator_stop > 0:
+                    break
+
+            batch = to_device(batch, "cuda" if ngpu > 0 else "cpu")
+            if no_forward_run:
+                continue
+
+            for turn in ["generator", "discriminator"]:
+                forward_generator = turn == "generator"
```
Comment (takenori-y) [inline espnet2/train/gan_trainer.py:373]: This line is the same as line 176. It can be written as a function.

**Review by takenori-y — APPROVED** [review]
LGTM. I added minor comments.

File: espnet2/train/gan_trainer.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+        grad_clip_type = options.grad_clip_type
+        log_interval = options.log_interval
+        no_forward_run = options.no_forward_run
+        ngpu = options.ngpu
+        use_wandb = options.use_wandb
+        generator_first = options.generator_first
+        distributed = distributed_option.distributed
+
+        if accum_grad > 1:
+            raise ValueError("accum_grad must be 1 in GAN-based training.")
```
Comment (sw005320) [inline espnet2/train/gan_trainer.py:138]: It's better to mention why accum_grad must be 1 (as a comment).

**Review by sw005320 — COMMENTED** [review]
Sorry that I cannot fully understand the reporter's behavior.
Does this use a separate reporter for a discriminator and generator?

**Comment by kan-bayashi** [comment]
> Does this use a separate reporter for a discriminator and generator?

Yes, it reports like this:

```python
[test:0/4] 2021-08-21 12:39:58,447 (gan_trainer:328) INFO: 580epoch:train:451-500batch: iter_time=6.912e-05, discriminator_forward_time=0.564, discriminator_loss=2.139, discriminator_real_loss=1.095, discriminator_fake_loss=1.044, discriminator_backward_time=0.200, discriminator_optim_step_time=0.010, optim1_lr0=1.860e-04, discriminator_train_time=0.861, generator_forward_time=0.053, generator_loss=36.109, generator_mel_loss=20.158, generator_kl_loss=2.092, generator_dur_loss=1.750, generator_adv_loss=2.815, generator_feat_match_loss=9.295, generator_backward_time=0.356, generator_optim_step_time=0.081, optim0_lr0=1.860e-04, generator_train_time=0.583, train_time=1.477
[test:0/4] 2021-08-21 12:40:42,256 (trainer:328) INFO: 580epoch results: [train] iter_time=0.002, discriminator_forward_time=0.566, discriminator_loss=2.148, discriminator_real_loss=1.099, discriminator_fake_loss=1.049, discriminator_backward_time=0.201, discriminator_optim_step_time=0.010, optim1_lr0=1.860e-04, discriminator_train_time=0.862, generator_forward_time=0.052, generator_loss=36.249, generator_mel_loss=20.202, generator_kl_loss=2.141, generator_dur_loss=1.747, generator_adv_loss=2.828, generator_feat_match_loss=9.331, generator_backward_time=0.356, generator_optim_step_time=0.082, optim0_lr0=1.860e-04, generator_train_time=0.584, train_time=1.511, time=12 minutes and 37.25 seconds, total_count=290000, gpu_max_cached_mem_GB=12.289, [valid] generator_loss=37.110, generator_mel_loss=19.741, generator_kl_loss=2.974, generator_dur_loss=2.144, generator_adv_loss=2.738, generator_feat_match_loss=9.513, discriminator_loss=2.115, discriminator_real_loss=1.021, discriminator_fake_loss=1.094, time=28.58 seconds, total_count=5220, gpu_max_cached_mem_GB=12.289, [att_plot] time=13.42 seconds, total_count=0, gpu_max_cached_mem_GB=12.289
```

スクリーンショット 2021-08-21 12 40 56

**Comment by sw005320** [comment]
I see, thanks!
This is very useful.

How about documenting it somewhere?
I think it would be good to make a document for the GAN class in general (in a follow-up PR?)
Then, it can be used for various purposes.

**Comment by sw005320** [comment]
Can you add a test?

**Comment by kan-bayashi** [comment]
I want to separate the documentation.
There is no test for trainer since it will be performed in integration test.
I added it #3449.

**Comment by sw005320** [comment]
> I want to separate the documentation since it can reduce the test time.
> There is no test for trainer since it will be performed in integration test.
> I added it #3449.

Make sense.
Thanks!

**Review by sw005320 — APPROVED** [review]

File: espnet2/train/abs_gan_espnet_model.py (hunk 2 above)
Comment (kan-bayashi) [reply espnet2/train/abs_gan_espnet_model.py:48]: I will add docstrings.

File: espnet2/train/gan_trainer.py (hunk 3 above)
Comment (kan-bayashi) [reply espnet2/train/gan_trainer.py:105]: Thank you for your suggestion.
This is nice but the current abstract model assumes that two way forward.
So for now I will keep the current style.

File: espnet2/train/gan_trainer.py (hunk 1 above)
Comment (kan-bayashi) [reply espnet2/train/gan_trainer.py:405]: I came up with idea.
Maybe inherit of `TrainerOptions` will solve this issue.

File: espnet2/train/abs_gan_espnet_model.py (hunk 2 above)
Comment (kan-bayashi) [reply espnet2/train/abs_gan_espnet_model.py:48]: Done.

File: espnet2/train/gan_trainer.py (hunk 5 above)
Comment (kan-bayashi) [reply espnet2/train/gan_trainer.py]: changed to one liner.

File: espnet2/train/gan_trainer.py (hunk 6 above)
Comment (kan-bayashi) [reply espnet2/train/gan_trainer.py]: Changed message.

File: espnet2/train/gan_trainer.py (hunk 4 above)
Comment (kan-bayashi) [reply espnet2/train/gan_trainer.py]: Moved and modified.

File: espnet2/train/abs_gan_espnet_model.py (hunk 2 above)
Comment (kan-bayashi) [reply espnet2/train/abs_gan_espnet_model.py:48]: added.

File: espnet2/train/gan_trainer.py (hunk 1 above)
Comment (kan-bayashi) [reply espnet2/train/gan_trainer.py]: Removed this part.
